libs/spock-shaded/src/main/groovy/org/codice/spock/extension/builtin/SupplementalExtension.groovy

```diff
@@ -0,0 +1,258 @@
+      (SortedMap): { new TreeMap<>() },
+      (NavigableMap): { new TreeMap<>() },
+      (Optional): Optional.empty(),
+      (BigInteger): BigInteger.ZERO,
+      (BigDecimal): BigDecimal.ZERO
+  ]
+
+  private def mockFactory = new DetachedMockFactory()
+
+  @Override
+  void visitSpecAnnotation(Supplemental annotation, SpecInfo spec) {
+    // register the {@link ConfineMetaClassChanges} interceptor first to restore all metaclass changes
+    // done by our own interceptor
+    spec.bottomSpec.addInterceptor(new ConfineMetaClassChangesInterceptor([Class, Method, System, MockingApi]))
+
+    spec.bottomSpec.addInterceptor(new IMethodInterceptor() {
+      @Override
+      void intercept(IMethodInvocation invocation) throws Throwable {
+        def ext = SupplementalExtension.this
+
+        // supplement the Class class
+        Class.metaClass {
+          getProxyableMethods { ext.getProxyableMethods(delegate) }
+          getMethodBySimplePrototype { String p -> ext.getMethodBySimplePrototype(delegate, p) }
+          getNoSpockSimpleName { ext.getNoSpockSimpleName(delegate) }
+        }
+        // supplement the Method class
+        Method.metaClass {
+          getSimplePrototype { ext.getSimplePrototype(delegate) }
+          verifyInvocation { MockInvocation i, Object... p -> ext.verifyInvocation(delegate, i, p) }
+        }
+        // supplement the System class
+        System.metaClass.static.setPropertyIfNotNull = ext.&setPropertyIfNotNull
+        // supplement specification class to allow creating dummies
+        invocation.instance.class.metaClass {
+          Dummy = ext.&createDummy
+          Dummies = ext.&createDummies
+        }
+        invocation.proceed()
+      }
+    })
+  }
+
+  /**
+   * Gets all proxy-able public methods (inherited or not) for the given class filtering away all
+   * final methods and all all methods defined by the {@link Object} class (e.g. {@link Object#equals},
+   * {@link Object#toString}, {@link Object#clone} ...).
+   *
+   * @param type the type for which to retrieve all proxy-able methods
+   * @return a list of all proxy-able methods for the given class
+   */
+  private def getProxyableMethods(Class<?> type) {
+    type.methods.findAll {
+      Modifier.isPublic(it.modifiers) && !Modifier.isFinal(it.modifiers) && (it.declaringClass != Object.class)
+    }
+  }
+
+  /**
+   * Returns a {@code Method} object that reflects the specified public member method of the class
+   * or interface represented by the specified class that matches the given simple prototype.
+   *
+   * @param type the type for which to retrieve a method matching the given prototype
+   * @param prototype the prototype to match
+   * @return the corresponding method
+   * @throws NoSuchMethodException if a matching method is not found
+   */
+  private def getMethodBySimplePrototype(Class<?> type, String prototype) {
+    def methods = type.methods.findAll { getSimplePrototype(it) == prototype }
```

Comment:
✏️ I recommend stripping out all whitespace from the passed in `prototype` to avoid false misses. 

I'm not sure where you're using this supplemental method so it might not matter, of course.

Reply:
Really good idea.

It is used as a quick way to get a method by name and types. Useful when the methods are overloaded. 

I originally added the simple prototype notion for making the test name much easier to read as I can then have something like:
`test that #method.simplePrototype fails when xxxx` which gets translated to something like `test that add(SolrInputDocument,Collection) fails when xxxx` which was good enough and needed to be a property of the Method class for it to be usable in the expansion done by `@Unroll`. Then I realized that it was making my life easier in places where I needed to get or remove a specific method without all the complications to get to that method with exact parameters.